crates/matrix-sdk-base/src/client.rs

```diff
@@ -1186,23 +1206,24 @@ impl BaseClient {
             new_rooms.invite.insert(room_id, new_info);
         }
 
-        // TODO remove this, we're processing account data events here again
+        // We're processing direct state events here separately
         // because we want to have the push rules in place before we process
         // rooms and their events, but we want to create the rooms before we
         // process the `m.direct` account data event.
-        let has_new_direct_room_data = response.account_data.events.iter().any(|raw_event| {
-            raw_event
-                .deserialize()
-                .map(|event| event.event_type() == GlobalAccountDataEventType::Direct)
-                .unwrap_or_default()
-        });
+        let has_new_direct_room_data = global_account_data_events
+            .iter()
+            .any(|event| event.event_type() == GlobalAccountDataEventType::Direct);
         if has_new_direct_room_data {
-            self.handle_account_data(&response.account_data.events, &mut changes).await;
+            self.process_direct_rooms(global_account_data_events, &mut changes).await;
```

Comment:
Don't pass the `Vec` but a reference to it, so basically a slice (because [a `Vec` derefs to a `slice`](https://doc.rust-lang.org/std/vec/struct.Vec.html#deref-methods-%5BT%5D))

```suggestion
            self.process_direct_rooms(&global_account_data_events, &mut changes).await;
```